talisker/requests.py

```diff
@@ -88,33 +91,90 @@ def send(request, **kwargs):
+            return func(request, **kwargs)
+        except Exception as e:
+            metadata = collect_metadata(request, None)
+            metadata['exception'] = str(e)
+            logger.exception('http request failure', extra=metadata)
+            raven.breadcrumbs.record(
+                type='http',
+                category='requests',
+                data=metadata,
+            )
+            raise
+
     send._inject_request_id = True
     return send
 
 
-def enable_metric_path_len(func):
+def enable_metric_control(func):
     @functools.wraps(func)
     def request(method, url, **kwargs):
         try:
             _local.metric_path_len = kwargs.pop('metric_path_len', 0)
+            _local.emit_metric = kwargs.pop('emit_metric', True)
             return func(method, url, **kwargs)
         finally:
             del _local.metric_path_len
-    request._metric_path_len = True
+            del _local.emit_metric
+    request._metric_control = True
     return request
 
 
+def collect_metadata(request, response):
+    metadata = collections.OrderedDict()
+    metadata['url'] = request.url
+    metadata['method'] = request.method
+
+    if response is not None:
+        metadata['status_code'] = response.status_code
+        duration = response.elapsed.total_seconds() * 1000
+        metadata['duration'] = duration
+
+    request_type = request.headers.get('content-type', None)
+    if request_type is not None:
+        metadata['request_type'] = request_type
+
+    try:
+        metadata['request_size'] = int(
+            request.headers.get('content-length', 0))
+    except ValueError:
+        pass
+
+    if response is not None:
+        response_type = response.headers.get('content-type', None)
+        if response_type is not None:
+            metadata['response_type'] = response_type
+        try:
+            metadata['response_size'] = int(
+                response.headers.get('content-length', 0))
+        except ValueError:
+            pass
+
+    return metadata
+
+
 def metrics_response_hook(response, **kwargs):
-    """Response hook that records statsd metrics"""
-    path_len = getattr(_local, 'metric_path_len', 0)
-    prefix, duration = get_timing(response, path_len)
-    statsd.get_client().timing(prefix, duration)
+    """Response hook that records statsd metrics and breadcrumbs."""
+    try:
+        metadata = collect_metadata(response.request, response)
+        logger.info('http request', extra=metadata)
+
+        # massage breadcrumb data to meet sentry expectations for http request
+        raven.breadcrumbs.record(
+            type='http', category='requests', data=metadata)
+
+        if getattr(_local, 'emit_metric', True):
+            path_len = getattr(_local, 'metric_path_len', 0)
+            metric_name = get_metric_name(response, path_len)
+            statsd.get_client().timing(metric_name, metadata['duration'])
+    except Exception:
+        logging.exception('response hook error')
```

Comment:
Does this need to `raise` too?

Reply:
No, that's the point of adding the try, we don't want metrics failures to break user code